File: frontend/src/components/dashboard/freeform/lib/formattingTypes.ts
// Plan 10a — TS mirror of backend/vizql/formatting_types.py.
// Build_Tableau.md §XIV.1 precedence: Mark > Field > Worksheet > DS > Workbook.

export enum StyleProp {
  FontFamily = 'font-family',
  FontSize = 'font-size',
  FontWeight = 'font-weight',
  FontStyle = 'font-style',
  Color = 'color',
  BackgroundColor = 'background-color',
  TextDecoration = 'text-decoration',
  TextAlign = 'text-align',
  LineHeight = 'line-height',
  NumberFormat = 'number-format',
  DateFormat = 'date-format',
  BorderTop = 'border-top',
  BorderRight = 'border-right',
  BorderBottom = 'border-bottom',
  BorderLeft = 'border-left',
  Padding = 'padding',
  ShowColumnBanding = 'show-column-banding',
  ShowRowBanding = 'show-row-banding',
  AxisTickColor = 'axis-tick-color',
  ZeroLineColor = 'zero-line-color',
  PaneLineThickness = 'pane-line-thickness',
}

export type SelectorKind = 'mark' | 'field' | 'sheet' | 'ds' | 'workbook';

export type Selector =
  | { kind: 'mark'; markId: string }
  | { kind: 'field'; fieldId: string }
  | { kind: 'sheet'; sheetId: string }
  | { kind: 'ds'; dsId: string }
  | { kind: 'workbook' };

export type StyleValue = string | number | boolean;

export interface StyleRule {
  readonly selector: Selector;
  readonly properties: Readonly<Record<StyleProp, StyleValue>>;
}

export interface LineStyle {
  weight: number;
  color: string;
  dash: 'solid' | 'dashed' | 'dotted' | 'dash-dot';
}

export const DEFAULT_LINE_STYLE: LineStyle = { weight: 1, color: '#000000', dash: 'solid' };

export interface StyledBox {
  backgroundColor: string;
  backgroundOpacity: number;
  borderTop: LineStyle;
  borderRight: LineStyle;
  borderBottom: LineStyle;
  borderLeft: LineStyle;
  shadow?: { x: number; y: number; blur: number; color: string };
}

export const DEFAULT_STYLED_BOX: StyledBox = {
  backgroundColor: '#ffffff',
  backgroundOpacity: 1.0,
  borderTop: DEFAULT_LINE_STYLE,
  borderRight: DEFAULT_LINE_STYLE,
  borderBottom: DEFAULT_LINE_STYLE,
  borderLeft: DEFAULT_LINE_STYLE,
};

export const NUMERIC_STYLE_PROPS: ReadonlySet<StyleProp> = new Set([
  StyleProp.FontSize,
  StyleProp.LineHeight,
  StyleProp.Padding,
  StyleProp.PaneLineThickness,
]);

export const BOOL_STYLE_PROPS: ReadonlySet<StyleProp> = new Set([
  StyleProp.ShowColumnBanding,
  StyleProp.ShowRowBanding,
]);

File: backend/vizql/formatting_types.py
"""Typed format primitives for the Plan 10a precedence resolver.

References:
    - Build_Tableau.md §XIV.1 precedence chain (Mark > Field > Worksheet > DS > Workbook)
    - Build_Tableau.md §XIV.5 shading / borders / dividers
    - Build_Tableau.md §XIV.6 StyledBox + LineStyle
    - Appendix C: tabstylemodel / tabdocformatting
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from types import MappingProxyType
from typing import Mapping, Union


class StyleProp(str, Enum):
    FONT_FAMILY = "font-family"
    FONT_SIZE = "font-size"
    FONT_WEIGHT = "font-weight"
    FONT_STYLE = "font-style"
    COLOR = "color"
    BACKGROUND_COLOR = "background-color"
    TEXT_DECORATION = "text-decoration"
    TEXT_ALIGN = "text-align"
    LINE_HEIGHT = "line-height"
    NUMBER_FORMAT = "number-format"
    DATE_FORMAT = "date-format"
    BORDER_TOP = "border-top"
    BORDER_RIGHT = "border-right"
    BORDER_BOTTOM = "border-bottom"
    BORDER_LEFT = "border-left"
    PADDING = "padding"
    SHOW_COLUMN_BANDING = "show-column-banding"
    SHOW_ROW_BANDING = "show-row-banding"
    AXIS_TICK_COLOR = "axis-tick-color"
    ZERO_LINE_COLOR = "zero-line-color"
    PANE_LINE_THICKNESS = "pane-line-thickness"


# --- Selectors ----------------------------------------------------------

@dataclass(frozen=True)
class MarkSelector:
    mark_id: str
    kind: str = "mark"


@dataclass(frozen=True)
class FieldSelector:
    field_id: str
    kind: str = "field"


@dataclass(frozen=True)
class WorksheetSelector:
    sheet_id: str
    kind: str = "sheet"


@dataclass(frozen=True)
class DataSourceSelector:
    ds_id: str
    kind: str = "ds"


@dataclass(frozen=True)
class WorkbookSelector:
    kind: str = "workbook"


Selector = Union[
    MarkSelector, FieldSelector, WorksheetSelector, DataSourceSelector, WorkbookSelector
]


_SPECIFICITY = {
    "mark": 5,
    "field": 4,
    "sheet": 3,
    "ds": 2,
    "workbook": 1,
}


def selector_specificity(s: Selector) -> int:
    """Higher = more specific. Used by resolver walk order."""
    return _SPECIFICITY[s.kind]


# --- StyleRule ----------------------------------------------------------

@dataclass(frozen=True)
class StyleRule:
    selector: Selector
    properties: Mapping[StyleProp, object]

    def __post_init__(self) -> None:
        # Freeze to prevent downstream mutation from leaking into memoised results.
        object.__setattr__(self, "properties", MappingProxyType(dict(self.properties)))


# --- LineStyle / StyledBox / RichText ----------------------------------

@dataclass(frozen=True)
class LineStyle:
    weight: int = 1
    color: str = "#000000"
    dash: str = "solid"  # "solid" | "dashed" | "dotted" | "dash-dot"


DEFAULT_LINE_STYLE = LineStyle()


@dataclass(frozen=True)
class Shadow:
    x: int = 0
    y: int = 0
    blur: int = 0
    color: str = "#00000000"


@dataclass(frozen=True)
class StyledBox:
    background_color: str = "#ffffff"
    background_opacity: float = 1.0
    border_top: LineStyle = field(default_factory=LineStyle)
    border_right: LineStyle = field(default_factory=LineStyle)
    border_bottom: LineStyle = field(default_factory=LineStyle)
    border_left: LineStyle = field(default_factory=LineStyle)
    shadow: Shadow | None = None


DEFAULT_STYLED_BOX = StyledBox()


@dataclass(frozen=True)
class RichTextRun:
    text: str
    style: Mapping[StyleProp, object] = field(default_factory=dict)

    def __post_init__(self) -> None:
        object.__setattr__(self, "style", MappingProxyType(dict(self.style)))


@dataclass(frozen=True)
class RichText:
    runs: tuple[RichTextRun, ...] = ()
